lib/core/objects.toit

```diff
@@ -187,24 +187,91 @@ class LazyInitializer_:
     // The __invoke_initializer__ builtin does a tail call to the method with the given id.
     return __invoke_initializer__ id_or_task_
 
+  initializing -> Task_:
+    if id_or_task_ is Task_: return id_or_task_
+    if id_or_task_ is InitializingTask_: return (id_or_task_ as InitializingTask_).task
+    throw "BAD_LAZY_INITIALIZER_CALL"
+
+  add_waiting task/Task_:
+    if id_or_task_ is Task_: id_or_task_ = InitializingTask_ id_or_task_
+    if not id_or_task_ is InitializingTask_: throw "BAD_LAZY_INITIALIZER_CALL"
+    (id_or_task_ as InitializingTask_).add_waiting task
+
+  do_waiting [block]:
+    if id_or_task_ is Task_: return
+    if not id_or_task_ is InitializingTask_: throw "BAD_LAZY_INITIALIZER_CALL"
+    current := (id_or_task_ as InitializingTask_).waiting
+    while current:
+      block.call current.task
+      current = current.next
+
+
+class InitializingTask_:
+  task / Task_
+  waiting / WaitingTask_? := null
+  last_waiting / WaitingTask_? := null
+
+  constructor .task:
+
+  add_waiting task/Task_:
+    new_last := WaitingTask_ task
+    if not waiting:
+      waiting = new_last
+      last_waiting = waiting
+    else:
+      last_waiting.next = new_last
+      last_waiting = new_last
+
+class WaitingTask_:
+  task /Task_
+  next /WaitingTask_? := null
+  constructor .task:
+
 /**
 Runs the $initializer function for the given $global.
 */
 run_global_initializer_ global/int initializer/LazyInitializer_:
-  if initializer.id_or_task_ is not int:
-    // There is already an initialization in progress.
-    initialization_in_progress_failure_ global
-
-  __store_global_with_id__ global (LazyInitializer_ task)
-
-  // If the initializer fails, we store the original initializer back in
-  // the global. This means that it is possible to invoke a global that throws
-  // again.
-  result / any := initializer
-  try:
-    result = initializer.call
-    return result
-  finally:
-    // Either store the computed result, if the initializer succeeded, or
-    // store the original initializer if it failed.
-    __store_global_with_id__ global result
+  while true:
+    if initializer.id_or_task_ is not int:
+      // There is already an initialization in progress.
+      initializing_task := initializer.initializing
+      this_task := task
```

Comment:
Move this out one level and use when creating the RAM-allocated LazyInitializer?

Reply:
done.